Add ROUND function in druid-sql.

KazuhitoT opened the pull request:
Add `ROUND` function in druid-sql.
issue: https://github.com/apache/incubator-druid/issues/5644

- It can be used for doubles (return `double`) and integers (return `long`).
- Both `ROUND(x)` and `ROUND(x, y)` works like `TRUNC(x, y)`.
- This is implemented by replacing `Math.round()` with `Bigdecimal.setScale()` as following:
```java
protected ExprEval eval(ExprEval param, ExprEval scale)
{
  if (param.type() == ExprType.LONG) {
    Long value = BigDecimal.valueOf(param.asLong()).setScale(scale.asInt(), RoundingMode.HALF_UP).longValue();
    return ExprEval.of(value);
  } else {
    double value = BigDecimal.valueOf(param.asDouble()).setScale(scale.asInt(), RoundingMode.HALF_UP).doubleValue();
    return ExprEval.of(value);
  }
}
```

asdf2014 commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -441,18 +443,43 @@ protected ExprEval eval(double param)
+      if (param.type() == ExprType.LONG) {
+        Long value = BigDecimal.valueOf(param.asLong()).setScale(scale.asInt(), RoundingMode.HALF_UP).longValue();
+        return ExprEval.of(value);
+      } else {
```
We should consider `ExprType.STRING` situation here.

asdf2014 commented on sql/src/test/java/org/apache/druid/sql/calcite/expression/ExpressionsTest.java:
```diff
@@ -354,6 +355,54 @@ public void testTruncate()
+        DruidExpression.fromExpression("round(\"z\")"),
+        -2.0
+    );
+  }
```
We should add more abnormal test cases, such as `inputRef("s")`.

asdf2014 commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -441,18 +443,43 @@ protected ExprEval eval(double param)
     {
-      return ExprEval.of(Math.round(param));
+      if (param.type() == ExprType.LONG) {
+        Long value = BigDecimal.valueOf(param.asLong()).setScale(scale.asInt(), RoundingMode.HALF_UP).longValue();
```
`Long value` used an unnecessary type boxing here.

asdf2014 commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/RoundOperatorConversion.java:
```diff
@@ -0,0 +1,67 @@
+
+public class RoundOperatorConversion implements SqlOperatorConversion
+{
+  private static final SqlFunction SQL_FUNCTION = OperatorConversions.operatorBuilder("Round")
```
We should use `OperatorConversions.operatorBuilder("ROUND")`.

asdf2014 commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/RoundOperatorConversion.java:
```diff
@@ -0,0 +1,67 @@
+public class RoundOperatorConversion implements SqlOperatorConversion
+{
+  private static final SqlFunction SQL_FUNCTION = OperatorConversions.operatorBuilder("Round")
+      .operandTypes(SqlTypeFamily.NUMERIC, SqlTypeFamily.INTEGER).requiredOperands(1)
```
For unifying the coding style, it would be better to write each function call of `OperatorConversions` on a different line.

asdf2014 commented:
@KazuhitoT BTW, we should also update the `math-expr.md` file at the same time.

KazuhitoT commented:
@asdf2014 
Thanks for reviewing my PR.🙏

Summary of changes are as follows:
- fixed code
- added abnormal test cases such as `ExprType.STRING`
- updated `math-expr.md`

Finally, `ROUND(x, y)` requires that
- x must be integer or double.
- y must be integer and can be negative, defaults to 0 if omitted.

asdf2014 commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -450,9 +452,43 @@ public String name()
+        }
+        return eval(value1, value2.asInt());
+      } else {
+        throw new IAE("Function[%s] needs 1 or 2 arguments", name());
```
We should check the size at the beginning of the apply method, then we can refactor the processing logic of `ExprEval value1` to eliminate duplicate code blocks.

asdf2014 commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -450,9 +452,43 @@ public String name()
+        ExprEval value1 = args.get(0).eval(bindings);
+        ExprEval value2 = args.get(1).eval(bindings);
+        if (value1.type() != ExprType.LONG && value1.type() != ExprType.DOUBLE) {
+          throw new IAE("first argument should be integer or double type but got %s type", value1.type());
```
It would be better to add the function name to this error message. For example, `The first argument to the function[%s] should be integer or double type but get the %s type`.

asdf2014 commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -450,9 +452,43 @@ public String name()
+          throw new IAE("first argument should be integer or double type but got %s type", value1.type());
+        }
+        if (value2.type() != ExprType.LONG) {
+          throw new IAE("second argument should be integer type but got %s type", value2.type());
```
It would be better to add the function name to this error message. For example, `The second argument to the function[%s] should be integer type but get the %s type`.

KazuhitoT commented:
@asdf2014 
Thanks for your reviewing.
I modified error messages and refactored codes.

gianm commented on docs/content/misc/math-expr.md:
```diff
@@ -133,7 +133,7 @@ See javadoc of java.lang.Math for detailed explanation for each function.
 |remainder|remainder(x, y) would return the remainder operation on two arguments as prescribed by the IEEE 754 standard|
 |rint|rint(x) would return value that is closest in value to x and is equal to a mathematical integer|
-|round|round(x) would return the closest long value to x, with ties rounding up|
+|round|round(x, y) would return the value of the x rounded to the y decimal places. While x can be an integer or floating-point number, y must be an integer. The type of the return value is specified by that of x. y defaults to 0 if omitted. When y is negative, x is rounded on the left side of the y decimal points.|
```
Please add this to the SQL docs too (querying/sql.md)

gianm commented on sql/src/main/java/org/apache/druid/sql/calcite/expression/builtin/RoundOperatorConversion.java:
```diff
@@ -0,0 +1,71 @@
+          ? inputExpressions.get(1).parse(plannerContext.getExprMacroTable())
+          : null;
+
+      if (digitsExpr == null) {
```
Not required, but, `DruidExpression.fromFunctionCall` might allow you to write this more cleanly.

gianm commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -508,9 +510,42 @@ public String name()
+      return eval(param, 0);
+    }
+
+    protected ExprEval eval(ExprEval param, int scale)
```
Could probably be private instead of protected, which if so, would be preferred.

gianm commented on core/src/main/java/org/apache/druid/math/expr/Function.java:
```diff
@@ -508,9 +510,42 @@ public String name()
+      }
+    }
+
+    protected ExprEval eval(ExprEval param)
```
Could probably be private instead of protected, which if so, would be preferred.

KazuhitoT commented:
@gianm
Thanks for your reviewing.

Summary of changes are as follows:
- Replaced `protected` with `private`.
- Updated `sql.md`.
- Used `DruidExpression.fromFunctionCall`.
- Modified tests, replaced `ClassCast Exception` with `IAE Exception`.

Travis job failed but it seems to be caused by some server errors. 
(My Travis job works: https://travis-ci.org/KazuhitoT/incubator-druid)
So, could you restart it and confirm if it's ok?

gianm commented:
Sure, I just restarted it.

KazuhitoT commented:
@gianm
Thanks. Travis job works.
Could you please merge this PR if it's ok?
